### PR TITLE
Adding multipart copy

### LINKED ISSUE #106: Investigate multipart copy
http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPartCopy.html

### PR DESCRIPTION
**Pull request by nhibberd**
Fixes #106

### CODE REVIEW

**Comment by nhibberd** [comment]
As per the [copy part request factory](https://github.com/aws/aws-sdk-java/blob/105b4d2f155737404b27e65e58b67dba11c18e19/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/internal/CopyPartRequestFactory.java#L73)  the last byte is defined `offset + size - 1` /cc @jystic factories to the rescue

**Comment by nhibberd** [comment]
initial perf ~ 5-6 x aws-cli/1.9.7

File: mismi-s3/src/Mismi/S3/Commands.hs [hunk 1]
```diff
@@ -169,19 +169,83 @@ read' a = do
...
+
+  case r of
+    Left f ->
+      abort >>
+        fail ("multipart copy failed in worker: " <> show f)
+
+    Right prts -> do
+      -- Sort is required here because the completeMultipartUpload api expects an
+      -- ascending list of part id's
+      let z = sortOn (\(PartResponse i _) -> i) $ snd prts
```
Comment (charleso) [inline mismi-s3/src/Mismi/S3/Commands.hs]: Worth sharing this whole block with `multipartUpload`, they look identical to me?

Comment (nhibberd) [inline mismi-s3/src/Mismi/S3/Commands.hs]: Yup

File: mismi-s3/main/s3.hs [hunk 2]
```diff
@@ -97,10 +96,15 @@ run c = do
     Size a ->
       getSize a >>= liftIO . maybe exitFailure (putStrLn . pack . show)
     Sync s d m f ->
-      eitherT (\se -> liftIO $ (hPutStrLn stderr $ renderSyncError se) >> exitFailure) pure (syncWithMode m s d f)
+      renderExit renderSyncError $ syncWithMode m s d f
     List a rq ->
       rec (list' a) (listRecursively' a) rq $$ DC.mapM_ (liftIO . putStrLn . addressToText)
 
+renderExit :: MonadIO m => (e -> Text) -> EitherT e m a -> m a
```
Comment (charleso) [inline mismi-s3/main/s3.hs]: Isn't this basically `orDie`?

Comment (charleso) [inline mismi-s3/main/s3.hs]: Or we don't use `x` in `mismi` perhaps?

Comment (nhibberd) [inline mismi-s3/main/s3.hs]: We do, we do

File: mismi-s3/src/Mismi/S3/Commands.hs [hunk 3]
```diff
@@ -175,25 +171,87 @@ read' a = do
...
+        (left $ MultipartCopyError f)
+
+    Right prts ->
+      flip onException abort $
+        void . send $ fencode' A.completeMultipartUpload dest mpu &
+          cMultipartUpload .~ pure (completedMultipartUpload & cmuParts .~ sortPartResponse (snd prts))
+
+-- Sort is required here because the completeMultipartUpload api expects an
+-- ascending list of part id's
+sortPartResponse :: [PartResponse] -> Maybe (NEL.NonEmpty A.CompletedPart)
```
Comment (charleso) [inline mismi-s3/src/Mismi/S3/Commands.hs]: Ta

**Comment by nhibberd** [comment]
Will probably push this throw then do another pass to ensure all the error types are accurate

File: mismi-s3/test/Test/IO/Mismi/S3/Commands.hs [hunk 4]
```diff
@@ -134,13 +144,46 @@ prop_copy_fail t = testAWS $ do
...
+
+  uploadOrFail s a
+  liftIO . putStrLn $ "Uploaded file"
+
+  liftIO . putStrLn $ "Running copy ..."
+  eitherT (fail . T.unpack . renderCopyError) pure $ copy a b
+
+  liftIO . putStrLn $ "Done copy"
+  -- compare
+  eitherT (fail . show) pure $ download b d
```
Comment (charleso) [inline mismi-s3/test/Test/IO/Mismi/S3/Commands.hs]: I guess using `read'` here is more annoying?

Comment (nhibberd) [inline mismi-s3/test/Test/IO/Mismi/S3/Commands.hs]: > 10GB

**Comment by charleso** [comment]
Surface :+1:

**Pull request merged**